### PR TITLE
Fix upload for existing on-demand content

### PR DESCRIPTION
**Pull request by gerrod3**
fixes: #5501

### CODE REVIEW

File: pulpcore/app/serializers/content.py [hunk 1]
```diff
@@ -64,16 +64,24 @@ def create(self, validated_data):
             content.touch()
         else:
             try:
-                with transaction.atomic():
-                    content = self.Meta.model.objects.create(**validated_data)
-                    for relative_path, artifact in artifacts.items():
-                        models.ContentArtifact.objects.create(
-                            artifact=artifact, content=content, relative_path=relative_path
-                        )
+                content = self.Meta.model.objects.create(**validated_data)
```
Comment (mdellweg) [inline pulpcore/app/serializers/content.py:67]: I think, this was actually correct. If we had created the content, we need to create the content_artifacts also. And we need to do so in a transaction for the content to never be incomplete.

File: pulpcore/app/serializers/content.py [hunk 2]
```diff
@@ -64,16 +64,24 @@ def create(self, validated_data):
...
+            cas = [
+                models.ContentArtifact(artifact=a, content=content, relative_path=rp)
+                for rp, a in artifacts.items()
+            ]
+            models.ContentArtifact.objects.bulk_create(
+                cas,
+                update_conflicts=True,
+                update_fields=["artifact"],
+                unique_fields=["content", "relative_path"],
+            )
```
Comment (mdellweg) [inline pulpcore/app/serializers/content.py:84]: What if you moved this block into the except branch? That should be the only codepath where the content could have unresolved remote artifacts.

**Review by mdellweg — CHANGES_REQUESTED** [review]
This issue gets me thinking:
What do we do, if someone uploads a file into another content that magically happend to match this content_artifact?
Don't we need to check that the artifacts we attach here match the expectations of the remote_artifacts for the content_artifact, like size and checksums?

File: pulpcore/app/serializers/content.py (hunk 2 above)
Comment (gerrod3) [reply pulpcore/app/serializers/content.py:84]: No, that's wrong. The except part will only be reached if the result of the first `retrieve` is `None`.

**Comment by mdellweg** [comment]
> > This issue gets me thinking: What do we do, if someone uploads a file into another content that magically happend to match this content_artifact? Don't we need to check that the artifacts we attach here match the expectations of the remote_artifacts for the content_artifact, like size and checksums?
> 
> I don't understand what you are trying to say. `ContentArtifact` is unique on `content` and `relative_path`. When you call `retrieve` you have the **validated** data of the artifact the user uploaded and their supplied metadata that the plugin's create api requires. This is meant so that you are guaranteed to pull the correct content unit if it already exists in Pulp by supplying the correct bits. This is only wrong if your implementation of `validate` is wrong and doesn't check that the user is supplying what they say they are supplying.

OK, I think I'm saying two loosely related things:
Let's ignore the first (upload of an artifact for a remote artifact of another content unit not know to the user).
But the second one is: If there is on demand content, it has content artifacts and hopefully remote artifacts attached. And as far as i know, these can contain checksums the actual artifact should meet. Now the user uploads artifacts that presumably create that content and we take it for that what if the checksums mismatch?
The thing is with on-demand content we kind of rely on the upstream to never lie to us. I start to think this is a really bad idea. Maybe remote content and content should not be mixed that way.

**Comment by gerrod3** [comment]
> OK, I think I'm saying two loosely related things: Let's ignore the first (upload of an artifact for a remote artifact of another content unit not know to the user). But the second one is: If there is on demand content, it has content artifacts and hopefully remote artifacts attached. And as far as i know, these can contain checksums the actual artifact should meet. Now the user uploads artifacts that presumably create that content and we take it for that what if the checksums mismatch? The thing is with on-demand content we kind of rely on the upstream to never lie to us. I start to think this is a really bad idea. Maybe remote content and content should not be mixed that way.

Yes, that is correct. Your content unit if it supports on-demand syncing should understand how to differentiate between content units from its metadata, either from the upstream or from an uploaded artifact. If the upstream lies about it's metadata then on upload you should create a different content unit, not re-use the bad content. This is a contract we expect our plugins to follow (which is why pulp_ansible is a mess since it doesn't follow it), so I am still not seeing your problem.

File: pulpcore/app/serializers/content.py (hunk 1 above)
Comment (dralley) [reply pulpcore/app/serializers/content.py:67]: Agree with this

File: pulpcore/tests/functional/api/using_plugin/test_content_access.py [hunk 3]
```diff
@@ -40,3 +40,32 @@ def test_file_remote_on_demand(
...
+    file_fixtures_root,
+    file_remote_factory,
+    file_repo,
+    file_bindings,
+    monitor_task,
+):
+    """Test that on-demand content can be uploaded to Pulp and become immediate content."""
+    remote = file_remote_factory(basic_manifest_path, policy="on_demand")
+    body = {"remote": remote.pulp_href}
+    monitor_task(file_bindings.RepositoriesFileApi.sync(file_repo.pulp_href, body).task)
```
Comment (dralley) [inline pulpcore/tests/functional/api/using_plugin/test_content_access.py:57]: Just reminding myself here - this doesn't pollute the shared fixture, does it?

Comment (gerrod3) [reply pulpcore/tests/functional/api/using_plugin/test_content_access.py:57]: No, all the pulp_file fixtures are unique per test. Each one is generated during the test, I don't think any test uses a shared fixtures.

Comment (dralley) [reply pulpcore/tests/functional/api/using_plugin/test_content_access.py:57]: Cool, the way pytest fixtures work is a bit unintuitive sometimes

**Comment by mdellweg** [comment]
Well, the one thing I am going to require is, that when we create a new content, we need to create the content artifacts in the same transaction.

Otherwise we open up a window for race conditions, where we have incorrect content in the database.

**Comment by gerrod3** [comment]
> Well, the one thing I am going to require is, that when we create a new content, we need to create the content artifacts in the same transaction.
> 
> Otherwise we open up a window for race conditions, where we have incorrect content in the database.

Ok, I reverted the create logic so now this is only a new check we perform when the content already existed in Pulp.

**Review by mdellweg — APPROVED** [review]

File: pulpcore/app/serializers/content.py [hunk 4]
```diff
@@ -74,6 +75,22 @@ def create(self, validated_data):
...
+        if not created:
+            ca_set = content.contentartifact_set
+            if ca_set.count() == 0 or ca_set.filter(artifact=None).exists():
+                cas = [
+                    models.ContentArtifact(artifact=a, content=content, relative_path=rp)
+                    for rp, a in artifacts.items()
+                ]
+                models.ContentArtifact.objects.bulk_create(
+                    cas,
+                    update_conflicts=True,
```
Comment (mdellweg) [inline pulpcore/app/serializers/content.py:90]: Is this the thing that basically turns this into an update for all the not yet associated artifacts?

**Review by mdellweg — APPROVED** [review]

**Review by ggainey — APPROVED** [review]

**Pull request merged**